### PR TITLE
Added Heap Sort in Python 3 and updated README.md

### LINKED ISSUE #627: Heap Sort in Python3
This is an open issue for the Heap Sort algorithm implementation in Python3. If you wish to contribute, just make some noise and we will assign you.

### PR DESCRIPTION
**Pull request by hasalfernando**
**Fixes #627**

By submitting this pull request I confirm I've read and complied with the below declarations.

-   [x] I have read the [Contribution guidelines](https://github.com/FOSS-UCSC/FOSSALGO/blob/master/CONTRIBUTING.md) 
and I am confident that my PR reflects them.

-   [x] I have followed the [coding guidelines](https://github.com/FOSS-UCSC/FOSSALGO/blob/master/CONTRIBUTING.md#cs) 
for this project.

-   [x] My code follows the [skeleton code structure](https://github.com/FOSS-UCSC/FOSSALGO/blob/master/CONTRIBUTING.md#sample).

-   [x] This pull request has a descriptive title. 
For example, `Added {Algorithm/DS name} [{Language}]`, not `Update README.md` or `Added new code`.

-   [x] This pull request shall only be merged once the build passes.

-   [x] I have mentioned the issue number correctly (with hyper-link) in this pull request description.

### CODE REVIEW

File: algorithms/ar-hsrt/python3/HeapSort.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    if r < n and arr[largest] < arr[r]: 
+        largest = r 
+  
+    if largest != i: 
+        arr[i],arr[largest] = arr[largest],arr[i] 
+  
+        heapify(arr, n, largest) 
+  
+def heapSort(arr): 
+    n = len(arr)
```
Comment (RavinduSachintha) [inline algorithms/ar-hsrt/python3/HeapSort.py:18]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6355988)

File: algorithms/ar-hsrt/python3/HeapSort.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
+def heapify(arr, n, i): 
+    largest = i
+    l = 2 * i + 1
+    r = 2 * i + 2
+  
+    if l < n and arr[i] < arr[l]: 
+        largest = l
```
Comment (RavinduSachintha) [inline algorithms/ar-hsrt/python3/HeapSort.py:7]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6355988)

File: algorithms/ar-hsrt/python3/HeapSort.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
+def heapify(arr, n, i): 
+    largest = i
+    l = 2 * i + 1
+    r = 2 * i + 2
+  
+    if l < n and arr[i] < arr[l]: 
+        largest = l 
+  
+    if r < n and arr[largest] < arr[r]:
```
Comment (RavinduSachintha) [inline algorithms/ar-hsrt/python3/HeapSort.py:9]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6355988)

File: algorithms/ar-hsrt/python3/HeapSort.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
+def heapify(arr, n, i): 
+    largest = i
+    l = 2 * i + 1
+    r = 2 * i + 2
+
```
Comment (RavinduSachintha) [inline algorithms/ar-hsrt/python3/HeapSort.py:5]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6355988)

File: algorithms/ar-hsrt/python3/HeapSort.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    l = 2 * i + 1
+    r = 2 * i + 2
+  
+    if l < n and arr[i] < arr[l]: 
+        largest = l 
+  
+    if r < n and arr[largest] < arr[r]: 
+        largest = r 
+  
+    if largest != i:
```
Comment (RavinduSachintha) [inline algorithms/ar-hsrt/python3/HeapSort.py:12]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6355988)

File: algorithms/ar-hsrt/python3/HeapSort.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  
+        heapify(arr, n, largest) 
+  
+def heapSort(arr): 
+    n = len(arr) 
+  
+    for i in range(n//2 - 1, -1, -1): 
+        heapify(arr, n, i) 
+  
+    for i in range(n-1, 0, -1):
```
Comment (RavinduSachintha) [inline algorithms/ar-hsrt/python3/HeapSort.py:23]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6355988)

File: algorithms/ar-hsrt/python3/HeapSort.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    n = len(arr) 
+  
+    for i in range(n//2 - 1, -1, -1): 
+        heapify(arr, n, i) 
+  
+    for i in range(n-1, 0, -1): 
+        arr[i], arr[0] = arr[0], arr[i]
+        heapify(arr, i, 0) 
+  
+arr = [15, 2, 8, 9, 0, 1, 12, 22]
```
Comment (RavinduSachintha) [inline algorithms/ar-hsrt/python3/HeapSort.py:27]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6355988)

File: algorithms/ar-hsrt/python3/HeapSort.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+  
+    if r < n and arr[largest] < arr[r]: 
+        largest = r 
+  
+    if largest != i: 
+        arr[i],arr[largest] = arr[largest],arr[i] 
+  
+        heapify(arr, n, largest) 
+  
+def heapSort(arr):
```
Comment (RavinduSachintha) [inline algorithms/ar-hsrt/python3/HeapSort.py:17]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6355988)

File: algorithms/ar-hsrt/python3/HeapSort.py [hunk 9]
```diff
@@ -0,0 +1,32 @@
+def heapify(arr, n, i): 
+    largest = i
+    l = 2 * i + 1
+    r = 2 * i + 2
+  
+    if l < n and arr[i] < arr[l]: 
+        largest = l 
+  
+    if r < n and arr[largest] < arr[r]: 
+        largest = r
```
Comment (RavinduSachintha) [inline algorithms/ar-hsrt/python3/HeapSort.py:10]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6355988)

File: algorithms/ar-hsrt/python3/HeapSort.py [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    if l < n and arr[i] < arr[l]: 
+        largest = l 
+  
+    if r < n and arr[largest] < arr[r]: 
+        largest = r 
+  
+    if largest != i: 
+        arr[i],arr[largest] = arr[largest],arr[i] 
+  
+        heapify(arr, n, largest)
```
Comment (RavinduSachintha) [inline algorithms/ar-hsrt/python3/HeapSort.py:15]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6355988)

File: algorithms/ar-hsrt/python3/HeapSort.py [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+    r = 2 * i + 2
+  
+    if l < n and arr[i] < arr[l]: 
+        largest = l 
+  
+    if r < n and arr[largest] < arr[r]: 
+        largest = r 
+  
+    if largest != i: 
+        arr[i],arr[largest] = arr[largest],arr[i]
```
Comment (RavinduSachintha) [inline algorithms/ar-hsrt/python3/HeapSort.py:13]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6355988)

File: algorithms/ar-hsrt/python3/HeapSort.py [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+  
+    for i in range(n//2 - 1, -1, -1): 
+        heapify(arr, n, i) 
+  
+    for i in range(n-1, 0, -1): 
+        arr[i], arr[0] = arr[0], arr[i]
+        heapify(arr, i, 0) 
+  
+arr = [15, 2, 8, 9, 0, 1, 12, 22] 
+heapSort(arr)
```
Comment (RavinduSachintha) [inline algorithms/ar-hsrt/python3/HeapSort.py:28]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6355988)

File: algorithms/ar-hsrt/python3/HeapSort.py [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+    if largest != i: 
+        arr[i],arr[largest] = arr[largest],arr[i] 
+  
+        heapify(arr, n, largest) 
+  
+def heapSort(arr): 
+    n = len(arr) 
+  
+    for i in range(n//2 - 1, -1, -1): 
+        heapify(arr, n, i)
```
Comment (RavinduSachintha) [inline algorithms/ar-hsrt/python3/HeapSort.py:21]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6355988)

File: algorithms/ar-hsrt/python3/HeapSort.py [hunk 14]
```diff
@@ -0,0 +1,32 @@
+def heapify(arr, n, i): 
+    largest = i
+    l = 2 * i + 1
+    r = 2 * i + 2
+  
+    if l < n and arr[i] < arr[l]:
```
Comment (RavinduSachintha) [inline algorithms/ar-hsrt/python3/HeapSort.py:6]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6355988)

File: algorithms/ar-hsrt/python3/HeapSort.py [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+  
+def heapSort(arr): 
+    n = len(arr) 
+  
+    for i in range(n//2 - 1, -1, -1): 
+        heapify(arr, n, i) 
+  
+    for i in range(n-1, 0, -1): 
+        arr[i], arr[0] = arr[0], arr[i]
+        heapify(arr, i, 0)
```
Comment (RavinduSachintha) [inline algorithms/ar-hsrt/python3/HeapSort.py:25]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6355988)

File: algorithms/ar-hsrt/python3/HeapSort.py [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+        heapify(arr, n, i) 
+  
+    for i in range(n-1, 0, -1): 
+        arr[i], arr[0] = arr[0], arr[i]
+        heapify(arr, i, 0) 
+  
+arr = [15, 2, 8, 9, 0, 1, 12, 22] 
+heapSort(arr) 
+n = len(arr) 
+print ("Sorted array is")
```
Comment (RavinduSachintha) [inline algorithms/ar-hsrt/python3/HeapSort.py:30]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6355988)

File: algorithms/ar-hsrt/python3/HeapSort.py [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+  
+    for i in range(n-1, 0, -1): 
+        arr[i], arr[0] = arr[0], arr[i]
+        heapify(arr, i, 0) 
+  
+arr = [15, 2, 8, 9, 0, 1, 12, 22] 
+heapSort(arr) 
+n = len(arr) 
+print ("Sorted array is") 
+for i in range(n):
```
Comment (RavinduSachintha) [inline algorithms/ar-hsrt/python3/HeapSort.py:31]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6355988)

File: algorithms/ar-hsrt/python3/HeapSort.py [hunk 18]
```diff
@@ -0,0 +1,32 @@
+def heapify(arr, n, i):
```
Comment (RavinduSachintha) [inline algorithms/ar-hsrt/python3/HeapSort.py:1]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6355988)

File: algorithms/ar-hsrt/python3/HeapSort.py [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+  
+    if largest != i: 
+        arr[i],arr[largest] = arr[largest],arr[i] 
+  
+        heapify(arr, n, largest) 
+  
+def heapSort(arr): 
+    n = len(arr) 
+  
+    for i in range(n//2 - 1, -1, -1):
```
Comment (RavinduSachintha) [inline algorithms/ar-hsrt/python3/HeapSort.py:20]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6355988)

File: algorithms/ar-hsrt/python3/heap_sort.py [hunk 20]
```diff
@@ -0,0 +1,32 @@
+def heapify(arr, n, i):
```
Comment (RavinduSachintha) [inline algorithms/ar-hsrt/python3/heap_sort.py:1]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6355988)

File: algorithms/ar-hsrt/python3/heap_sort.py [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+    if l < n and arr[i] < arr[l]:
+        largest = l
+
+    if r < n and arr[largest] < arr[r]:
+        largest = r
+
+    if largest != i: 
+        arr[i],arr[largest] = arr[largest],arr[i]
+        heapify(arr, n, largest)
+
```
Comment (RavinduSachintha) [inline algorithms/ar-hsrt/python3/heap_sort.py:15]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6355988)

File: algorithms/ar-hsrt/python3/heap_sort.py [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+    l = 2 * i + 1
+    r = 2 * i + 2
+
+    if l < n and arr[i] < arr[l]:
+        largest = l
+
+    if r < n and arr[largest] < arr[r]:
+        largest = r
+
+    if largest != i:
```
Comment (RavinduSachintha) [inline algorithms/ar-hsrt/python3/heap_sort.py:12]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/gh/FOSS-UCSC/FOSSALGO/pullRequest?prid=6355988)

**Review by RavinduSachintha — APPROVED** [review]

**Comment by RavinduSachintha** [comment]
@all-contributors please add @hasalfernando for code and doc

**Comment by RavinduSachintha** [comment]
@hasalfernando thank you for your contribution. keep it up :+1: 
make sure to add a star to this repo :wink:

**Comment by hasalfernando** [comment]
Always happy to contribute. 🙌❤️

**Pull request merged**